middlewares/tracing/tracing.go

```diff
@@ -0,0 +1,204 @@
+		return fmt.Sprintf("entrypoint %s %s", t.ep, r.Host)
+	}
+
+	ctx, _ := t.Extract(opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(r.Header))
+	span := t.StartSpan(opNameFunc(r), ext.RPCServerOption(ctx))
+	ext.Component.Set(span, t.ServiceName)
+	ext.HTTPMethod.Set(span, r.Method)
+	ext.HTTPUrl.Set(span, r.URL.String())
+	span.SetTag("http.host", r.Host)
+	ext.SpanKindRPCServer.Set(span)
+
+	w = &statusCodeTracker{w, 200}
+	r = r.WithContext(opentracing.ContextWithSpan(r.Context(), span))
+
+	next(w, r)
+
+	code := uint16(w.(*statusCodeTracker).status)
+	ext.HTTPStatusCode.Set(span, code)
+	if code >= 400 {
+		ext.Error.Set(span, true)
+	}
+	span.Finish()
+}
+
+type fwdMiddleware struct {
+	frontend string
+	backend  string
+	opName   string
+	*Tracing
+}
+
+// NewForwarder creates a new forwarder middleware that the final outgoing request
+func (t *Tracing) NewForwarder(frontend, backend string) negroni.Handler {
+	return &fwdMiddleware{
+		Tracing:  t,
+		frontend: frontend,
+		backend:  backend,
+		opName:   fmt.Sprintf("forward %s/%s", frontend, backend),
+	}
+}
+
+func (t *fwdMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	ctx := r.Context()
+	span, _ := opentracing.StartSpanFromContext(ctx, t.opName)
+	span.SetTag("frontend.name", t.frontend)
+	span.SetTag("backend.name", t.backend)
+	ext.HTTPMethod.Set(span, r.Method)
+	ext.HTTPUrl.Set(span, r.URL.String())
+	span.SetTag("http.host", r.Host)
+	ext.SpanKindRPCClient.Set(span)
+
+	opentracing.GlobalTracer().Inject(
+		span.Context(),
+		opentracing.HTTPHeaders,
+		opentracing.HTTPHeadersCarrier(r.Header))
+
+	w = &statusCodeTracker{w, 200}
+	r = r.WithContext(opentracing.ContextWithSpan(ctx, span))
+
+	next(w, r)
+
+	code := uint16(w.(*statusCodeTracker).status)
+	ext.HTTPStatusCode.Set(span, code)
+	if code >= 400 {
+		ext.Error.Set(span, true)
+	}
+	span.Finish()
+}
+
+// LogEventf logs an event to the span in the request context.
+func LogEventf(r *http.Request, str string, vals ...interface{}) {
+	if span := opentracing.SpanFromContext(r.Context()); span != nil {
+		span.LogEvent(fmt.Sprintf(str, vals...))
+	}
+}
+
+// LogFields logs the opentracing log fields to the span in the request context.
+func LogFields(r *http.Request, flds ...otlog.Field) {
+	if span := opentracing.SpanFromContext(r.Context()); span != nil {
+		span.LogFields(flds...)
+	}
+}
+
+// StartSpan starts a new span from the one in the request context
+func StartSpan(r *http.Request, operationName string, opts ...opentracing.StartSpanOption) (opentracing.Span, *http.Request) {
+	span, ctx := opentracing.StartSpanFromContext(r.Context(), operationName, opts...)
+	r = r.WithContext(ctx)
+	return span, r
+}
+
+type baseMiddleware struct {
+	opName string
+	*Tracing
+}
+
+// NewSpanMiddleware creates a new middleware wraps a span around subsequent
+// middleware
+func (t *Tracing) NewSpanMiddleware(name string) negroni.Handler {
```

Comment:
`NewSpanMiddleware` seems to not be used. What is his purpose ?

Reply:
I intended it as the main means for adding tracing to new middlewares, however that didn't pan out. I'll remove it.